torch/_decomp/decompositions.py

```diff
@@ -1509,6 +1509,7 @@ def native_batch_norm(
 # In two weeks or so, we should remove this decomposition and phase out the current native_batch_norm
 # to be _native_batch_norm_legit and have the right schema (stating that there are input mutations).
 @aten.native_batch_norm.default.py_impl(DispatchKey.Autograd)
+@aten.native_batch_norm.default.py_impl(DispatchKey.CompositeImplicitAutograd)
```

Comment:
This is a superset of the key above no?
Also the comment above needs updating?

Reply:
the comment definitely needs updating 😛

So, the slightly-contentious idea behind this PR is that lower in the stack, I updated `torch._decomps.get_decomps()` to check **not only** the decomp table to see if a given op has a decomposition available, but to also check the dispatcher for any decomps registered to the `CompositeImplicitAutograd` key (link: https://github.com/pytorch/pytorch/pull/105865/files#diff-7008e894af47c01ee6b8eb94996363bd6c5a43a061a2c13a472a2f8a9242ad43R190)

There's one problem though: we don't actually make any hard guarantees that a given key in the dispatcher points does or does not point to a decomposition. We do rely pretty heavily, however, on the fact that everything registered to the `CompositeImplicitAutograd` key is in fact a decomposition into other ops.

QAT would like this API to faithfully return "the set of all decomps that would have run if we had traced through the dispatcher". However, `native_batch_norm` is an example of an op that has a pre-autograd decomp registered to it (through `op.py_impl()`, but the decomp is registered **directly** to the Autograd key instead of being registered to the CompositeImplicitAutograd key.

If we want to provide a guarantee to QAT that they can programatically access all decomps that would have run during tracing, then we need to make sure that every decomp we register to the Autograd key is also registered to the CompositeImplicitAutograd key.

This might sound kind of painful (since it requires auditing), but I think in practice this basically only applies to `native_batch_norm`.